### DISCUSSION TITLE
Debugging why smart card works in Chromium and Firefox, but not Chrome

### DISCUSSION THREAD

**Discussion by jat255** [OpenSC/OpenSC#3133] [Q&A] [answered]
I'm not really sure where to start looking, but my Smart Card / PIV reader works in Chromium and Firefox (on a Linux 6.6.29-1-lts kernel), but I cannot get it to work in Google Chrome (124.0.6367.91), which is my primary browser.

I'm using a `Identiv SCR3500 A Contact Reader` USB reader, as reported by `pkcs11-tool -L`.

In Firefox 125.0.2, I get the following prompt when trying to sign in to various services with my smart card and it lets me login:

![image](https://github.com/OpenSC/OpenSC/assets/1278301/2abc7e7a-ac3d-4f90-a69f-239675d38348)

In Chromium 124.0.6367.91, it also works, but makes me select my certificate from the card first:

![image](https://github.com/OpenSC/OpenSC/assets/1278301/68a9a3d2-20e1-42e3-a3d3-1e3ebe123352)

![image](https://github.com/OpenSC/OpenSC/assets/1278301/354c4ec8-5673-4fdf-8369-e988bafc64c6)

In Chrome (124.0.6367.91), I'm never presented with a certificate list, and just see a message as follows:

![image](https://github.com/OpenSC/OpenSC/assets/1278301/d994767a-a425-4adf-93bc-ba271bd32038)

Likewise, the light on my card reader does not flash at all (like it does in Chromium and FF), making it seem like it's not event trying to communicate with the card reader.

Any ideas?

**Comment by dengert** [comment]
This might be helpful:
https://linuxkamarada.com/en/2019/09/26/setting-up-smart-card-authentication-on-google-chrome-chromium/

"Chrome for Linux manages digital certificates similarly to Firefox — using [Mozilla NSS](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS) as backend. But, unlike Firefox, Chrome does not provide a graphical user interface to install [PKCS11](https://en.wikipedia.org/wiki/PKCS_11) modules. So, to set up Chrome you need to use the command line."

What distro of Linux are you using?

**Comment by msetina** [comment]
It is imperative to add the distribution information. 
First thing is that Chrome and Chromium have a settings page for certificates (chrome://settings/certificates). If it is not there you need to fix NSS db. 
On the other hand if we assume ubuntu based (snap install of Chromium) I think there is a problem in snap containment (https://bugs.launchpad.net/ubuntu/+source/chromium-browser/+bug/1859643). I was wondering what the problem was on my 20.04 installs and now I found that bug info. I hope they made something for 24.04.  Plain debian does not show this problem.

**Reply by msetina** [reply]
If you can confirm that you problem is in snap based browser, then the problem is basically a flaw of thinking in snap confinement:
https://bugs.launchpad.net/ubuntu/+source/chromium-browser/+bug/1741074
This is open here:
https://forum.snapcraft.io/t/confined-browser-snaps-cant-use-system-libraries-pkcs11-and-native-host-messaging-what-do-we-do/11828

I hope we get some love from Canonical...

**Reply by msetina** [reply]
This is OT, but snap confinement allows  interfaces.
https://forum.snapcraft.io/t/interface-management/6154

**Comment by dengert** [comment]
Installing chrome from google downloads on Xubuntu 22.04 does not use snap. It installs  `/usr/bin/google-chrome-stable` "Version 124.0.6367.118 (Official Build) (64-bit)"


`nss-tools` was already installed, so in home directory running : 
`modutil -dbdir sql:.pki/nssdb/ -add "OpenSC" -libfile /opt/ossl-3.1.2/lib/pkcs11/opensc-pkcs11.so` where my OpenSC development  OpenSC pkcs11 module is located. 

This created ~/.pki/nssdb/pkcs11.txt:
```
library=
name=NSS Internal PKCS #11 Module
parameters=configdir='sql:.pki/nssdb' certPrefix='' keyPrefix='' secmod='' flags= updatedir='' updateCertPrefix='' updateKeyPrefix='' updateid='' updateTokenDescription='' 
NSS=Flags=internal,critical trustOrder=75 cipherOrder=100 slotParams=(1={slotFlags=[ECC,RSA,DSA,DH,RC2,RC4,DES,RANDOM,SHA1,MD5,MD2,SSL,TLS,AES,Camellia,SEED,SHA256,SHA512] askpw=any timeout=30})

library=/opt/ossl-3.1.2/lib/pkcs11/opensc-pkcs11.so
name=opt-3.1.2-OpenSC
```

With aa Idemia PIV test card in Idemia SCR3500 reader  and running `/usr/bin/google-chrome-stable`
does not show much, but clicking on `settings`  then in "search window" typing "cert" a pop up window  "sign in to security device" asking "Please sign in to  SDK TEST CARD 1 PIV Authentic...."
"Password:" So I did and the  OpenSC debug log shows it logged it to PIV card. 

It does not show anything else but it looks like it will work with PIV.

I also use FireFox-esr which does not use snap which works with OpenSC. It does have under setting  a way to look at certificates and loaded pkcs11 modules and it looks like it also used the `/opt/ossl-3.1.2/lib/pkcs11/opensc-pkcs11.so` from a : `~/.mozilla/firefox-esr/./41gzc0n2.default-esr91/pkcs11.txt` which also loads p11-kit pkcs11 module..

Chrome also picked up a `Nitro Start` token I have been using too and asked to login to the token.

**Comment by jat255** [comment, marked as answer]
Sorry all for the delay as I was out of the office. This is on an Arch Linux machine, so not using snap.

I have previously run the `modutil -add` command, and I have the following entry in my `pkcs11.txt` file:

```
library=/usr/lib/opensc-pkcs11.so
name=OpenSC Smart Card Module
```

Unlike @daringer, I do not see any such "Sign in to security device" in my Chrome settings. Here's what I see when I search for "cert":

![image](https://github.com/OpenSC/OpenSC/assets/1278301/20edd81a-52fb-46d8-9cf6-d8ce88e5134e)

And the security page shows:

![image](https://github.com/OpenSC/OpenSC/assets/1278301/83d29249-982e-4014-a53d-09649d79aa6d)

I clicked on "Manage certificates", and nothing was shown, other than an option to browse to import a certificate.

On an off chance, I decided to completely close Google Chrome and re-open it, and oddly it was now working (without any changes). I guess maybe this is some sort of time-out hardware communication bug with Chrome? I often leave my browser open for many days through sleep/wake cycles, so perhaps it's related to that. Sorry for the noise and bother in that case. I'll close this out.

**Comment by driftwoodlk** [comment]
Somewhat-related issue ... I have my certs working in Firefox, but not Chrome/Edge.
_certutil -L -d .pki/nssdb/ -h "all"_ seems fine.
_modutil -dbdir sql:$HOME/.pki/nssdb/ -list_ shows the "OpenSC smartcard framework (0.22)"

But I primarily followed https://github.com/M-Pepper/linux-cac-walkthrough (I'm on Pop!_OS, but I don't think it matters - did installs through Synaptic), and it directs to load **onepin-opensc-pkcs11.so** instead of **opensc-pkcs11.so**.  I can't find any documentation on what the onepin version is.  Any advice?  I suppose I will delete the module and unload the security device in Firefox, and then re-do with the non-onepin version.

**Reply by dengert** [reply]
@driftwoodlk 
"OpenSC smartcard framework (0.22)" is from 2021 Current version is 0.26 from 11/2024 https://github.com/opensc/opensc/releases

Can you try with newer version?

Google for: opensc onepin

Some cards have more than one user pin, for example each key could have its own pin, or one to unlock the card and one to sign. OpenPGP for example. PKCS11 C_Login() does not have any support for more than one user pin per slot.  So to support cards with more than one pin, OpenSC defines multiple slots for the same card, one for each user pin. The number of slots and "virtual slots: can be set in  `opensc.conf`. But some pkcs11 applications are not flexible enough to use multiple slots, so the `onepin-opensc-pkcs11.so` module was added. 

So pick the one that works for your card. 

The "onepin" problem goes back to 2003.  `git log | grep onepin | wc` shows 33 entries.

**Reply by driftwoodlk** [reply]
I figured it out late last night - due to Flatpak issues #4066 and #4723, my Flatpak versions of Chrome/Edge weren't getting OpenSC pass through.  I installed non-Flatpak versions, and it worked fine!  (Now my issue is that Microsoft Remote Desktop Web Client apparently doesn't pass the smart card through into AVD, but that's another discussion.)
